sei-protocol/sei-cosmos
# Introduce Expedited Proposals

## Problem Statement
Currently the long voting window is used to provide opportunities for users to review the proposals and have enough time to vote appropriately to express their opinions. While this is the expected and desirable behavior, it is sometimes not ideal, for example:
1. When the whole system needs to respond immediately to an ongoing threats or a chain needs to enact an emergency upgrade via a text proposal, but needs to bypass the standard voting period.
2. There are some cases where a gov proposal needs to be passed to mitigate an outage, and if voting windows are really long (say a week), the impact of outage could cause disasters
3. Today when a severe vulnerability is uncovered the validator community can discuss an upcoming update in private and prepare an update but it is difficult to do this in a way that does not leak advanced warning for someone to prepare an attack against the network while this is occurring. As the value of assets increase further it is conceivable that even one of the individuals associated with an existing validator in these private channels may be tempted to leverage information to mount an attack before a mitigation is in place.


## Proposed Concept
This change introduces a new option to create an expedited proposal for emergency purpose. If a proposal is expedited, it uses a shorter **voting period** and a higher **tally threshold** and requires to reach a higher **quorum**. If the higher threshold is not met under the short expedited voting period, the expedited proposal is then converted as if it was a regular proposal and tallied / voted on under regular rules.

The distinction is achieved by having an IsExpedited flag on the proposal struct. When a proposal is converted from expedited to regular, this flag is flipped and VotingEndTime is extended. The rest of the fields, including the ProposalId remain the same. As a result, all existing votes are preserved.

The expedited governance proposal option would add a capability for a network to respond rapidly during emergency situations while balancing protection of rights against abuse where a proposal is pushed through quickly without time for proper debate. This can be achieved by providing an additional governance controlled set of parameters that establish a much higher level of approval such as a 67% majority or similar value appropriately chosen by the network.

Similar concept has been proposed and achieved [here](https://wallet.keplr.app/chains/osmosis/proposals/278)

## Changes
- add **IsExpedited** flag to protos and some proposal content struct 
- add **ExpeditedVotingPeriod** to protos of gov voting parameters (Default: 1 day)
- add **ExpeditedThreshold** to protos of gov tally parameters (Default: 2/3)
- add **ExpeditedQuorum** to protos of gov tally parameters (Default: 2/3)
- add **MinExpeditedDeposit** flag to indicate the minimum deposit needed for an expedited proposal to enter voting period
- implement changes in abci of x/gov to move expedited proposals to regular once voting period ends with proposal not meeting ExpeditedThreshold
- update documentation for proto change

**Supporting Proposal:**
- Text Proposal
- Parameter Change Proposal
- Upgrade Proposal

## Testing performed to validate your change
Added unit and integration tests to cover all new logic related to expedited proposals

Integration Test Locally:
1. Start a chain
2. Submit an expedited proposal
3. Make sure Voting End Time should be calculated based on expedited voting period
4. Vote Yes before expedited voting period
5. Proposal should pass
6. Submit another expedited proposal
7. Do not vote and let it expire
8. Make sure Voting End Time is updated based on regular voting period
